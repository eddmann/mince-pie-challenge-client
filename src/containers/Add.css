.Add .Form {
  max-width: 450px !important;
}
