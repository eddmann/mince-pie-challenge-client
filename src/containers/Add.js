import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Button, Form, Grid, Header, Message, Segment } from 'semantic-ui-react';
import { addPie, uploadPhoto } from '../actions/pie';

import './Add.css';

export class Add extends Component {
  static propTypes = {
    addPie: PropTypes.func.isRequired,
    uploadPhoto: PropTypes.func.isRequired,
    isSubmitting: PropTypes.bool.isRequired,
    error: PropTypes.string,
    addUrl: PropTypes.string.isRequired,
    pendingPieId: PropTypes.string,
    photoRequestUrl: PropTypes.string,
  };

  state = {
    name: '',
    photo: undefined,
  };

  onAdd = e => {
    e.preventDefault();
    const { name } = this.state;
    this.props.addPie(this.props.addUrl, name);
  };

  onPhotoSelection = e => {
    this.setState({ photo: e.target.files[0] });
  };

  onPhotoUpload = e => {
    e.preventDefault();
    this.props.uploadPhoto(this.props.pendingPieId, this.props.photoRequestUrl, this.state.photo);
  };

  onFormUpdate = e => this.setState({ [e.target.id]: e.target.value });

  renderAddForm = () => {
    return (
      <Form size="large">
        <Segment stacked>
          <Form.Input
            fluid
            id="name"
            icon="user"
            iconPosition="left"
            placeholder="Name"
            value={this.state.name}
            onChange={this.onFormUpdate}
          />
          <Button loading={this.props.isSubmitting} fluid color="blue" size="large" onClick={this.onAdd}>
            Add
          </Button>
        </Segment>
      </Form>
    );
  };

  renderPhotoUploadForm = () => {
    return (
      <Form size="large">
        <Segment stacked>
          <input type="file" onChange={this.onPhotoSelection} />
          <Button
            disabled={!this.state.photo}
            loading={this.props.isSubmitting}
            fluid
            color="blue"
            size="large"
            onClick={this.onPhotoUpload}
          >
            Upload
          </Button>
        </Segment>
      </Form>
    );
  };

  render() {
    return (
      <div className="Add">
        <Grid textAlign="center">
          <Grid.Column className="Form">
            <Header as="h2" color="teal" textAlign="center">
              {this.props.photoRequestUrl ? 'Upload Photo' : 'Add Pie'}
            </Header>
            {this.props.error && <Message error>{this.props.error}</Message>}
            {this.props.photoRequestUrl ? this.renderPhotoUploadForm() : this.renderAddForm()}
          </Grid.Column>
        </Grid>
      </div>
    );
  }
}

const mapStateToProps = ({ urls, add, photo }) => ({
  isSubmitting: add.isAdding || photo.isProcessing,
  error: add.error || photo.error,
  addUrl: urls.addUrl,
  pendingPieId: photo.pendingPieId,
  photoRequestUrl: photo.photoRequestUrl,
});

export default connect(mapStateToProps, { addPie, uploadPhoto })(Add);
