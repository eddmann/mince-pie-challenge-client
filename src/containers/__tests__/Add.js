import React from 'react';
import { Add } from '../Add';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

const ID = '1234';
const NAME = 'Mince Pie';
const PHOTO = { name: 'pie.jpg', type: 'image/jpeg', extension: 'jpg' };
const ADD_URL = 'add.url';
const REQUEST_URL = 'request.url';
const UPLOAD_URL = 'upload.url';

it('add the mince pies details', () => {
  const addPie = jest.fn();
  const wrapper = shallow(<Add
    addPie={addPie}
    uploadPhoto={jest.fn()}
    isSubmitting={false}
    error={undefined}
    addUrl={ADD_URL}
    pendingPieId={undefined}
    photoRequestUrl={undefined}
  />);

  wrapper.find('FormInput').simulate('change', { target: { id: 'name', value: NAME }});
  wrapper.find('Button').simulate('click', { preventDefault: jest.fn() });

  expect(addPie).toBeCalledWith(ADD_URL, NAME);
});

it('uploads a mince pies photo', () => {
  const uploadPhoto = jest.fn();
  const wrapper = shallow(<Add
    addPie={jest.fn()}
    uploadPhoto={uploadPhoto}
    isSubmitting={false}
    error={undefined}
    addUrl={ADD_URL}
    pendingPieId={ID}
    photoRequestUrl={REQUEST_URL}
  />);

  wrapper.find('input').simulate('change', { target: { files: [ PHOTO ] } });
  wrapper.find('Button').simulate('click', { preventDefault: jest.fn() });

  expect(uploadPhoto).toBeCalledWith(ID, REQUEST_URL, PHOTO);
});

it('displays add errors', () => {
  const wrapper = shallow(<Add
    addPie={jest.fn()}
    uploadPhoto={jest.fn()}
    isSubmitting={false}
    error={'Error'}
    addUrl={ADD_URL}
    pendingPieId={ID}
    photoRequestUrl={REQUEST_URL}
  />);

  expect(wrapper.find('Message')).toMatchSnapshot();
});
