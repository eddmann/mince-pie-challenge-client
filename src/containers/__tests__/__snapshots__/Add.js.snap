// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays add errors 1`] = `
<Message
  error={true}
>
  Error
</Message>
`;
